src/dbnode/storage/index/block.go

```diff
@@ -406,54 +396,149 @@ func (b *block) Query(
 	results BaseResults,
 	logFields []opentracinglog.Field,
 ) (bool, error) {
-	ctx, sp := ctx.StartTraceSpan(tracepoint.BlockQuery)
-	sp.LogFields(logFields...)
+	ctx, sp, sampled := ctx.StartSampledTraceSpan(tracepoint.BlockQuery)
 	defer sp.Finish()
+	if sampled {
+		sp.LogFields(logFields...)
+	}
 
-	exhaustive, err := b.queryWithSpan(ctx, cancellable, query, opts, results, sp, logFields)
+	exhaustive, err := b.queryNoLock(ctx, cancellable, query, opts, results)
 	if err != nil {
 		sp.LogFields(opentracinglog.Error(err))
 	}
 
 	return exhaustive, err
 }
 
-func (b *block) queryWithSpan(
+const (
+	queryGroupReadersParallelism = 8
+	queryGroupSize               = 8
+)
+
+type queryGroup struct {
+	readers    []m3ninxindex.Reader
+	exhaustive bool
+	err        error
+}
+
+func (b *block) segmentReadersNoLock() ([]segment.Reader, error) {
```

Comment:
nit Can the naming scheme for this style of function be `segmentReadersLocking`? A bit jarring seeing a lock as the first statement of the `NoLock` function